3p/integration.js

```diff
@@ -45,15 +45,26 @@ register('twitter', twitter);
 
 /**
  * Visible for testing.
- * Draws an ad to the window. Expects the data to include the ad type.
+ * Draws a 3p embed to the window. Expects the data to include the 3p type.
  * @param {!Window} win
  * @param {!Object} data
+ * @param {function(!Object, function())|undefined} configCallback
+ *     Optional callback that allows user code to manipulate the incoming
+ *     configuration. See
+ *     https://github.com/ampproject/amphtml/issues/1210 for some context
+ *     on this.
  */
-export function draw3p(win, data) {
+export function draw3p(win, data, configCallback) {
   const type = data.type;
-  assert(window.context.location.originValidated != null,
+  assert(win.context.location.originValidated != null,
       'Origin should have been validated');
-  run(type, win, data);
+  if (configCallback) {
+    configCallback(data, () => {
```

Comment:
Yah, I tend not to grok "mutate this object" APIs. Either passing the data (or returning it) makes things much more clear, but we can't support async calls with the return value.